daemon/device_unix.go

```diff
@@ -0,0 +1,81 @@
+//go:build !windows
+// +build !windows
+
+package daemon
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/containerd/pkg/userns"
+	"github.com/opencontainers/runc/libcontainer/devices"
+)
+
+// Testing dependencies
+var (
+	ioutilReadDir         = ioutil.ReadDir
+	usernsRunningInUserNS = userns.RunningInUserNS
+)
+
+// HostDevices returns all devices that can be found under /dev directory.
+// Based on HostDevices from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix.go#L71-L74
+func HostDevices() ([]*devices.Device, error) {
+	return GetDevices("/dev")
+}
+
+// GetDevices recursively traverses a directory specified by path
+// and returns all devices found there.
+// Based on GetDevices from runc:
+// https://github.com/opencontainers/runc/blob/v1.0.0/libcontainer/devices/device_unix.go#L76-L120
+func GetDevices(path string) ([]*devices.Device, error) {
+	files, err := ioutilReadDir(path)
+	if err != nil {
+		if errors.Is(err, os.ErrPermission) && usernsRunningInUserNS() {
+			// ignore the "permission denied" error if running in userns
+			return nil, nil
+		}
```

Comment:
So, I'm not "super" happy with us having to maintain a copy of this function. With my other suggestions applied, I went looking what the difference was between the upstream version in runc, and it looks like this is the only difference (compared to https://github.com/opencontainers/runc/blob/55162941720ba91b87c7d05f9e6a91b9f308f33d/libcontainer/devices/device_unix.go#L76-L120);

[rootless_version.txt](https://github.com/moby/moby/files/7632861/rootless_version.txt)
[runc_version.txt](https://github.com/moby/moby/files/7632862/runc_version.txt)


```patch
diff --git a/runc_version.txt b/rootless_version.txt
index 21ead00a57..7dcd6df5c8 100644
--- a/runc_version.txt
+++ b/rootless_version.txt
@@ -1,9 +1,13 @@
-func GetDevices(path string) ([]*Device, error) {
+func GetDevices(path string) ([]*devices.Device, error) {
 	files, err := osReadDir(path)
 	if err != nil {
+		if errors.Is(err, os.ErrPermission) && usernsRunningInUserNS() {
+			// ignore the "permission denied" error if running in userns
+			return nil, nil
+		}
 		return nil, err
 	}
-	var out []*Device
+	var out []*devices.Device
 	for _, f := range files {
 		switch {
 		case f.IsDir():
@@ -24,9 +28,9 @@ func GetDevices(path string) ([]*Device, error) {
 		case f.Name() == "console":
 			continue
 		}
-		device, err := DeviceFromPath(filepath.Join(path, f.Name()), "rwm")
+		device, err := devices.DeviceFromPath(filepath.Join(path, f.Name()), "rwm")
 		if err != nil {
-			if errors.Is(err, ErrNotADevice) {
+			if errors.Is(err, devices.ErrNotADevice) {
 				continue
 			}
 			if os.IsNotExist(err) {
@@ -34,7 +38,7 @@ func GetDevices(path string) ([]*Device, error) {
 			}
 			return nil, err
 		}
-		if device.Type == FifoDevice {
+		if device.Type == devices.FifoDevice {
 			continue
 		}
 		out = append(out, device)
```

To make things worse, containerd _also_ keeps a copy of this exact function (with some slight changes). I'm concerned that these implementations will start to diverge more, and we'll be trying to catch up with any change made (the `ioutil` -> `os` switch already being a (although not "important") example.


So, I'm wondering if this is a change that would be acceptable for upstream. Given that the runc maintainers prefer to keep `libcontainer` "internal" (reduce the number of external uses), perhaps we should use the containerd implementation instead. If we use the containerd variant, we also wouldn't have to convert the type to `oci.Device`;

https://github.com/moby/moby/blob/30413e5efb6c29da0df23dc95ae4625f31dd8d71/daemon/oci_linux.go#L882


Comparing the runc implementation with the  containerd implementation; [containerd_version.txt](https://github.com/moby/moby/files/7632956/containerd_version.txt) (expand the "details" section below)

<details>

```patch
diff --git a/runc_version.txt b/containerd_version.txt
index 21ead00a57..d195224696 100644
--- a/runc_version.txt
+++ b/containerd_version.txt
@@ -1,9 +1,25 @@
-func GetDevices(path string) ([]*Device, error) {
-	files, err := osReadDir(path)
+func getDevices(path, containerPath string) ([]specs.LinuxDevice, error) {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "error stating device path")
+	}
+
+	if !stat.IsDir() {
+		dev, err := deviceFromPath(path)
+		if err != nil {
+			return nil, err
+		}
+		if containerPath != "" {
+			dev.Path = containerPath
+		}
+		return []specs.LinuxDevice{*dev}, nil
+	}
+
+	files, err := ioutil.ReadDir(path)
 	if err != nil {
 		return nil, err
 	}
-	var out []*Device
+	var out []specs.LinuxDevice
 	for _, f := range files {
 		switch {
 		case f.IsDir():
@@ -13,7 +29,11 @@ func GetDevices(path string) ([]*Device, error) {
 			case "pts", "shm", "fd", "mqueue", ".lxc", ".lxd-mounts", ".udev":
 				continue
 			default:
-				sub, err := GetDevices(filepath.Join(path, f.Name()))
+				var cp string
+				if containerPath != "" {
+					cp = filepath.Join(containerPath, filepath.Base(f.Name()))
+				}
+				sub, err := getDevices(filepath.Join(path, f.Name()), cp)
 				if err != nil {
 					return nil, err
 				}
@@ -24,9 +44,9 @@ func GetDevices(path string) ([]*Device, error) {
 		case f.Name() == "console":
 			continue
 		}
-		device, err := DeviceFromPath(filepath.Join(path, f.Name()), "rwm")
+		device, err := deviceFromPath(filepath.Join(path, f.Name()))
 		if err != nil {
-			if errors.Is(err, ErrNotADevice) {
+			if err == errNotADevice {
 				continue
 			}
 			if os.IsNotExist(err) {
@@ -34,10 +54,10 @@ func GetDevices(path string) ([]*Device, error) {
 			}
 			return nil, err
 		}
-		if device.Type == FifoDevice {
-			continue
+		if containerPath != "" {
+			device.Path = filepath.Join(containerPath, filepath.Base(f.Name()))
 		}
-		out = append(out, device)
+		out = append(out, *device)
 	}
 	return out, nil
 }
```

</details>

We can ignore the `containerPath` argument (it's not used for `HostDevices()`, and we have our own logic for this)

The containerd version looks to be missing the part that skips FIFO devices;

```go
if device.Type == FifoDevice {
	continue
}
```

Which was added in https://github.com/opencontainers/runc//commit/95a59bf206f86bc449c4ccc22e2c08e78c707eb6 (https://github.com/opencontainers/runc/pull/2529), so I'm wondering if that's an omission in containerd @AkihiroSuda  ?

Reply:
I opened https://github.com/containerd/containerd/pull/6306 for that last bit

Reply:
That's true. Since we can't rely on `libcontainer`, if you think that `containerd` should be the upstream, let's move it there.

Reply:
I was giving it a quick try; let me push what I have as a PR there (but feel free to pick it up from there)